pennylane/qinfo/fisher.py

```diff
@@ -0,0 +1,235 @@
+
+    >>> cfim_func = qml.qinfo.classical_fisher(circ)
+    >>> cfim_func(params)
+    tensor([[1., 1.],
+        [1., 1.]], requires_grad=True)
+
+    This new function accepts the same signature as the :class:`.QNode`. Here is a small example with multiple arguments:
+
+    .. code-block:: python
+
+        @qml.qnode(dev)
+        def circ(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=0)
+            return qml.probs(wires=range(n_wires))
+
+    >>> x, y = pnp.array([0.5, 0.6], requires_grad=True)
+    >>> circ(x, y)
+    (tensor([0.87380224, 0.        , 0.12619776, 0.        ], requires_grad=True)
+    >>> qml.qinfo.classical_fisher(circ)(x, y)
+     [tensor([[0.15828019]], requires_grad=True),
+      tensor([[0.74825326]], requires_grad=True)])
+
+    Note how in the case of multiple variables we get a list of matrices with sizes
+    ``[(n_params0, n_params0), (n_params1, n_params1)]``, which in this case is simply two ``(1, 1)`` matrices.
+
+
+    A typical setting where the classical fisher information matrix is used is in variational quantum algorithms.
+    Closely related to the `quantum natural gradient <https://arxiv.org/abs/1909.02108>`_, which employs the `quantum` fisher information matrix,
+    we can compute a rescaled gradient using the CFIM. In this scenario, typically a Hamiltonian objective function :math:`\langle H \rangle` is minimized:
+
+    .. code-block:: python
+
+        H = qml.Hamiltonian(coeffs = [0.5, 0.5], ops = [qml.PauliZ(0), qml.PauliZ(1)])
+
+        @qml.qnode(dev)
+        def circ(params):
+            qml.RX(params[0], wires=0)
+            qml.RY(params[1], wires=0)
+            qml.RX(params[2], wires=1)
+            qml.RY(params[3], wires=1)
+            qml.CNOT(wires=(0,1))
+            return qml.expval(H)
+
+        params = pnp.random.random(4)
+
+    We can compute both the gradient of :math:`\langle H \rangle` and the CFIM with the same :class:`.QNode` ``circ`` in this example since ``classical_fisher()`` ignores the return types
+    and assumes ``qml.probs()`` for all wires.
+
+    >>> grad = qml.grad(circ)(params)
+    >>> cfim = qml.qinfo.classical_fisher(circ)(params)
+    >>> print(grad.shape, cfim.shape)
+    (4,) (4, 4)
+
+    Combined together, we can get a rescaled gradient to be employed for optimization schemes like natural gradient descent.
+
+    >>> rescaled_grad = cfim @ grad
+    >>> print(rescaled_grad)
+    [-0.66772533 -0.16618756 -0.05865127 -0.06696078]
+
+    """
+    new_qnode = _make_probs(qnode, post_processing_fn=lambda x: qml.math.squeeze(qml.math.stack(x)))
+
+    interface = qnode.interface
+
+    if interface in ("jax", "jax-jit"):
+        import jax
+
+        jac = jax.jacobian(new_qnode, argnums=argnums)
+
+    if interface == "torch":
+        jac = _torch_jac(new_qnode)
+
+    if interface == "autograd":
+        jac = qml.jacobian(new_qnode)
+
+    if interface == "tf":
+        jac = _tf_jac(new_qnode)
```

Comment:
Question: do we always want to compute the Jacobian using backprop here? Do we want to change the behaviour if e.g. the user explicitly specifies `diff_method="parameter_shift"` in the QNode device?

Reply:
Very good point! At the moment it was intended as a simulation-only feature, but this should be considered as well!